Sources/SemanticIndex/TaskScheduler.swift

```diff
@@ -349,13 +351,66 @@ package actor TaskScheduler<TaskDescription: TaskDescriptionProtocol> {
   ///  - `.medium`: 4
   ///  - `.low`: 2
   ///  - `.background`: 2
-  private let maxConcurrentTasksByPriority: [(priority: TaskPriority, maxConcurrentTasks: Int)]
+  private var maxConcurrentTasksByPriority: [(priority: TaskPriority, maxConcurrentTasks: Int)] {
+    didSet {
+      // These preconditions need to match the ones in `init`
+      maxConcurrentTasksByPriority = maxConcurrentTasksByPriority.sorted(by: { $0.priority > $1.priority })
+      precondition(maxConcurrentTasksByPriority.allSatisfy { $0.maxConcurrentTasks >= 0 })
+      precondition(maxConcurrentTasksByPriority.map(\.maxConcurrentTasks).isSorted(descending: true))
+      precondition(!maxConcurrentTasksByPriority.isEmpty)
+
+      // Check we are over-subscribed in currently executing tasks. If we are, cancel currently executing task to be
+      // rescheduled until we are within the new limit.
+      var tasksToReschedule: [QueuedTask<TaskDescription>] = []
+      for (priority, maxConcurrentTasks) in maxConcurrentTasksByPriority {
+        var tasksInPrioritySlot = currentlyExecutingTasks.filter { $0.priority <= priority }
+        if tasksInPrioritySlot.count <= maxConcurrentTasks {
+          // We have enough available slots. Nothing to do.
+          continue
+        }
+        tasksInPrioritySlot = tasksInPrioritySlot.sorted { $0.priority > $1.priority }
+        while tasksInPrioritySlot.count > maxConcurrentTasks {
+          // Cancel the task with the lowest priority (because it is least important and also takes a slot in the lower
+          // priority execution buckets) and the among those the most recent one because it has probably made the least
+          // progress.
```

Comment:
Shouldn't this be "and also takes a slot in the higher priority execution buckets"? Also doesn't seem like the recency is checked?

Reply:
This made me think about the entire logic that determines if we can schedule a task again and I found that it was a little broken if we have more than one priority. The fix is a little bigger and more involved, so I opened a separate PR for it: https://github.com/swiftlang/sourcekit-lsp/pull/1991